code/modules/client/preferences.dm

```diff
@@ -492,3 +492,130 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+	for(var/data_key in raw_data)
+		if(character_slot_regex.Find(data_key))
+			var/slot_id = text2num(replacetext(data_key, "character", ""))
+			if(slot_id > max_save_slots)
+				continue
+
+			var/list/character_data_raw = raw_data[data_key]
+			load_character(slot_id)
+			for(var/character_data_key in character_data_raw)
+				if(handle_import_preference(character_data_key, character_data_raw[character_data_key], PREFERENCE_CHARACTER))
+					continue
+				stack_trace("unhandled character preference import raw data key '[character_data_key]'")
+			save_character()
+			continue
+
+		if(handle_import_preference(data_key, raw_data[data_key], PREFERENCE_PLAYER))
+			continue
+
+		if(handle_import_raw(data_key, raw_data[data_key]))
+			continue
+
+		stack_trace("unhandled preference import raw data key '[data_key]'")
+
+	// save and load to trigger full refresh on preferences
+	save_preferences()
+	load_preferences()
+
+/// Handles importing data that isnt managed using preferences
+/datum/preferences/proc/handle_import_raw(key, data)
+	// keys that should not be overwritten when imported
+	var/static/list/ignored_keys = list(
+		"muted",
+		"lastchangelog",
+		"hearted_until",
+		"version",
+	)
+
+	if(key in ignored_keys)
+		return TRUE
+
+	switch(key)
+		if("default_slot")
+			default_slot = sanitize_integer(data, 1, max_save_slots, initial(default_slot))
+			return TRUE
+
+		if("chat_toggles")
+			chat_toggles = sanitize_integer(data, NONE, ~NONE, initial(toggles))
+			return TRUE
+
+		if("toggles")
+			toggles = sanitize_integer(data, NONE, ~NONE, initial(toggles))
+			return TRUE
+
+		if("ignoring")
+			ignoring = SANITIZE_LIST(data)
+			return TRUE
+
+		if("be_special")
+			be_special = sanitize_be_special(SANITIZE_LIST(data))
+			return TRUE
+
+		if("favorite_outfits")
+			var/list/parsed_favs = list()
+			for(var/typetext in SANITIZE_LIST(data))
+				var/datum/outfit/path = text2path(typetext)
+				if(ispath(path)) //whatever typepath fails this check probably doesn't exist anymore
+					parsed_favs += path
+			favorite_outfits = unique_list(parsed_favs)
+			return TRUE
+
+		if("key_bindings")
+			key_bindings = sanitize_keybindings(SANITIZE_LIST(data))
+			key_bindings_by_key = get_key_bindings_by_key(key_bindings)
+			return TRUE
+
+/// Handles the import of a preference based on its savefile key. Returns true if the savefile key exists
+/datum/preferences/proc/handle_import_preference(key, data, expected_savefile_identifier)
+	if(!(key in GLOB.preference_entries_by_key))
+		return FALSE
+
+	var/datum/preference/prefence_instance = GLOB.preference_entries_by_key[key]
+	if(prefence_instance.savefile_identifier != expected_savefile_identifier)
+		stack_trace("attempted to write to an invalid savefile based on the identifier! ([prefence_instance.savefile_identifier]) != (expected_savefile_identifier)")
+	else if(!write_preference(prefence_instance, data))
+		stack_trace("failed to write preference key '[key]' with data '[data]'")
```

Comment:
Can this be reached with bad user input?

Reply:
yes, if the key matches it passes through to write_preference

Reply:
Principally we shouldn't stack trace for user input